posts/2024-05-07-ultimate-string-guide.md

```diff
@@ -0,0 +1,1250 @@
+    - [ByteString](https://hackage.haskell.org/package/bytestring-0.12.1.0/docs/Data-ByteString.html#t:ByteString) (strict and lazy)
+    - [ShortByteString](https://hackage.haskell.org/package/bytestring-0.12.1.0/docs/Data-ByteString-Short.html#t:ShortByteString)
+- bytes sequences dealing with platform API differences
+    - [OsString](https://hackage.haskell.org/package/os-string-2.0.2.1/docs/System-OsString.html#t:OsString)
+    - [PosixString](https://hackage.haskell.org/package/os-string-2.0.2.1/docs/System-OsString-Posix.html#t:PosixString)
+    - [WindowsString](https://hackage.haskell.org/package/os-string-2.0.2.1/docs/System-OsString-Windows.html#t:WindowsString)
+- FFI types
+    - [CString](https://hackage.haskell.org/package/base-4.19.1.0/docs/Foreign-C-String.html#t:CString)
+    - [CStringLen](https://hackage.haskell.org/package/base-4.19.1.0/docs/Foreign-C-String.html#t:CStringLen)
+- filepath types (just type synonyms)
+    - [FilePath](https://hackage.haskell.org/package/filepath-1.5.2.0/docs/System-FilePath.html#t:FilePath)
+    - [OsPath](https://hackage.haskell.org/package/filepath-1.5.2.0/docs/System-OsPath.html#t:OsPath)
+    - [PosixPath](https://hackage.haskell.org/package/filepath-1.5.2.0/docs/System-OsPath-Posix.html#t:PosixPath)
+    - [WindowsPath](https://hackage.haskell.org/package/filepath-1.5.2.0/docs/System-OsPath-Windows.html#t:WindowsPath)
+
+If we delve more into filepaths, there are actually even more, e.g. strongly typed filepaths. But those are out of scope.
+
+## Which String type to pick
+
+In this section, we will examine each string like type and what its properties and use cases are.
+`String` was already discussed and we don't recommend it, so it's omitted here.
+
+### Text
+
+If you are not sure what you need, you most likely want `Text` from the [text package](https://hackage.haskell.org/package/text),
+which is shipped with GHC. This type is meant for human readable Unicode text and has all the primitives you need. The
+API is in fact more complete than the one for `String`, containing functions like `stripPrefix` and `toLower`.
+
+Internally, Text uses a UTF-8 encoded byte array since version 2.0 and UTF-16 before version 2.0. So it is always
+guaranteed to be valid unicode.
+
+The current definition for strict `Text` is (as of [2.1.1](https://hackage.haskell.org/package/text-2.1.1)):
+
+```hs
+-- | A space efficient, packed, unboxed Unicode text type.
+data Text = Text
+    {-# UNPACK #-} !A.Array -- ^ bytearray encoded as UTF-8
+    {-# UNPACK #-} !Int     -- ^ offset in bytes (not in Char!), pointing to a start of UTF-8 sequence
+    {-# UNPACK #-} !Int     -- ^ length in bytes (not in Char!), pointing to an end of UTF-8 sequence
+```
+
+As we can see here, this type allows efficient slicing to avoid unnecessary `memcpy` for many operations.
+E.g. `init` and `tail` are *O(1)* time and space. `splitAt` is *O(1)*, but *O(n)* time, because UTF-8 complicates
+the offset computation.
```

Comment:
```suggestion
E.g. `init` and `tail` are *O(1)* time and space. `splitAt` is not *O(1)*, but *O(n)* time, because UTF-8 complicates
the offset computation.
```